File: week2/JavaScript/task1.js
function findAndPrint(messages, currentStation) {
  const stations = {
    'Songshan': 0,
    'Nanjing Sanmin': 1,
    'Taipei Arena': 2,
    'Nanjing Fuxing': 3,
    'Songjiang Nanjing': 4,
    'Zhongshan': 5,
    'Beimen': 6,
    'Ximen': 7,
    'Xiaonanmen': 8,
    'Chiang Kai-Shek Memorial Hall': 9,
    'Guting': 10,
    'Taipower Building': 11,
    'Gongguan': 12,
    'Wanlong': 13,
    'Jingmei': 14,
    'Dapinglin': 15,
    'Qizhang': 16,
    'Xindian City Hall': 17,
    'Xindian': 18,
    'Xiaobitan': 16,
  };
  let locations = {};
  for (let station in stations) {
    for (let name in messages) {
      if (messages[name].includes(station)) {
        locations[name] = stations[station];
      }
    }
  }
  const currentStationIndex = stations[currentStation];
  let distanceObj = {};
  for (let name in locations) {
    const d = Math.abs(locations[name] - currentStationIndex);
    if (
      currentStation === 'Xiaobitan' ||
      messages[name].includes('Xiaobitan')
    ) {
      distanceObj[name] = d + 1;
    } else {
      distanceObj[name] = d;
    }
  }
  let distanceArr = [];
  for (i in distanceObj) {
    distanceArr.push(distanceObj[i]);
  }
  for (key in distanceObj) {
    if (distanceObj[key] === Math.min(...distanceArr)) {
      console.log(key);
    }
  }
}
const messages = {
  Bob: "I'm at Ximen MRT station.",
  Mary: 'I have a drink near Jingmei MRT station.',
  Copper: 'I just saw a concert at Taipei Arena.',
  Leslie: "I'm at home near Xiaobitan station.",
  Vivian: "I'm at Xindian station waiting for you.",
};
console.log('==========TASK 1==========');
findAndPrint(messages, 'Wanlong'); // print Mary
findAndPrint(messages, 'Songshan'); // print Copper
findAndPrint(messages, 'Qizhang'); // print Leslie
findAndPrint(messages, 'Ximen'); // print Bob
findAndPrint(messages, 'Xindian City Hall'); // print Vivian

File: week2/Python/ task1.py
def find_and_print(messages, current_station):
    stations=["Songshan", "Nanjing Sanmin", "Taipei Arena", "Nanjing Fuxing", "Songjiang Nanjing", "Zhongshan", "Beimen", "Ximen", "Xiaonanmen", "Chiang Kai-Shek Memorial Hall", "Guting", "Taipower Building", "Gongguan", "Wanlong", "Jingmei", "Dapinglin", "Qizhang", "Xindian City Hall", "Xindian", "Xiaobitan"]
    stations_index_dic={}
    for station in stations:
        i=stations.index(station)
        stations_index_dic[station]=i
    stations_index_dic["Xiaobitan"]=stations_index_dic["Qizhang"]
    locations={}
    for station in stations:
        for name in messages:
            if station in messages[name]:
                locations[name]=stations_index_dic[station]
    index_of_current_station=stations_index_dic[current_station]
    distance_dic={}
    for name in locations:
        d = abs(locations[name]-index_of_current_station)
        if current_station=="Xiaobitan" or ("Xiaobitan" in messages[name]):
            distance_dic[name]=d+1
        else:
            distance_dic[name]=d
    print(min(distance_dic, key=distance_dic.get))

messages={
    "Leslie":"I'm at home near Xiaobitan station.", 
    "Bob":"I'm at Ximen MRT station.",
    "Mary":"I have a drink near Jingmei MRT station.", 
    "Copper":"I just saw a concert at Taipei Arena.", 
    "Vivian":"I'm at Xindian station waiting for you."
}
print("==========TASK 1==========")
find_and_print(messages, "Wanlong") # print Mary 
find_and_print(messages, "Songshan") # print Copper 
find_and_print(messages, "Qizhang") # print Leslie 
find_and_print(messages, "Ximen") # print Bob 
find_and_print(messages, "Xindian City Hall") # print Vivian
